### ISSUE TITLE
Autentification problem with Android 11

### ISSUE THREAD

**Issue by DariaLorens** [aws-amplify/amplify-android#948]
Hello! I encountered a problem while authenticating. When I use this method:

```kotlin
Amplify.Auth.signIn("email", "password",
    { result -> Log.i("AuthQuickstart", if (result.isSignInComplete) "Sign in succeeded" else "Sign in not complete") },
    { error -> Log.e("AuthQuickstart", error.toString()) }
)
```

This exception appears:
> E/UserContextDataProvider: Exception in creating JSON from context data

It is caused by the method:

```java
public String getEncodedContextData(Context context, String username, String userPoolId, String signatureSecret) {
    JSONObject jsonResponse = new JSONObject();

    try {
        final Map<String, String> contextData = aggregator.getAggregatedData(context);
        JSONObject payload = getJsonPayload(contextData, username, userPoolId);
        String payloadString = payload.toString();

        String signature = signatureGenerator.getSignature(payloadString, signatureSecret, VERSION);
        jsonResponse = getJsonResponse(payloadString, signature);
        return getEncodedResponse(jsonResponse);
    } catch (Exception e) {
        Log.e(TAG, "Exception in creating JSON from context data");
        return null;
    }
}
```

> Error: java.lang.SecurityException: getDataNetworkTypeForSubscriber

After analyzing, I found that the problem occurs on Android 11.  My
`targetSdkVersion` is 30. If the application requests
`permission.READ_PHONE_STATE`, the exception disappears. What am I doing wrong?
Thanks!

**Comment by jamesonwilliams** [comment]
Hi @DariaLorens!

Unfortunately there isn't enough info here right now to help. Perhaps you can help me to better understand your issue.

Is `getEncodedContextData(...)` a method in your application code base? I don't see you calling it from inside the sign-in callback. Where/how do you call it? How does it relate to your use of the Amplify library?

Secondly, your `try`-`catch` is very broad. Potentially, several different errors being caught by the broad `Exception` type that you use:
```
try {
    final Map<String, String> contextData = aggregator.getAggregatedData(context);
    JSONObject payload = getJsonPayload(contextData, username, userPoolId);
    String payloadString = payload.toString();

    String signature = signatureGenerator.getSignature(payloadString, signatureSecret, VERSION);
    jsonResponse = getJsonResponse(payloadString, signature);
    return getEncodedResponse(jsonResponse);
} catch (Exception e) {
    Log.e(TAG, "Exception in creating JSON from context data");
    return null;
}
```

To facilitate debugging, I recommend to move code out of the `try`-`catch`, and to use finer granularity in your `catch` statement. Additionally, you should include the exception in your log message, so you can follow the stack trace:
```java
} catch (SpecificException1 specificException1) {
    Log.e(TAG, "Exception doing specific thing 1.", specificException1);
} catch (SpecificException2 specificException2) {
    Log.e(TAG, "Exception doing specific thing 2.", specificException2);
}
```

**Comment by hugo-chq** [comment]
have also run into an issue on a Pixel 3 device recently updated to android 11,
But no issues in the emulator with 11...

**Comment by solariswu** [comment]
I got the same error message when using the Amplify SDK. 
```
    implementation 'com.amplifyframework:aws-auth-cognito:1.1.1'
```

E/UserContextDataProvider: Exception in creating JSON from context data

It may not affect the application usage if you are not using Cognito Advanced Security Feature. However, would suggest to have a fix to avoid this. 

The code pasted is from aws-android-sdk-cognitoidentityprovider-asf with some assumption. 

https://github.com/sandfox/aws-android-sdk-cognitoidentityprovider-asf/blob/eddf44e6ee47a13b7416b78a44e134111eaedb9e/com/amazonaws/cognito/clientcontext/data/UserContextDataProvider.java#L36

**Comment by jamesonwilliams** [comment]
Sounds like a few folks are encountering this issue, so I'm re-opening it.

**Comment by hugo-chq** [comment]
Update: 
Our issue ended up being the device clock having an incorrect time after updating to 11, working now.

**Comment by TrekSoft** [comment]
Thanks for that update @hugo-chq and yes per @solariswu the code in question is from the UserContextDataProvider class used by AWSMobileClient. 

Digging into this a bit, I don't see how it's possible for that code to result in the `getDataNetworkTypeForSubscriber` SecurityException referenced by the original filer. For anyone seeing that exception, is it possible you are using the Analytics category as well or have some code set to execute when the user logs in that could be causing this?

Also version 1.1.1 is quite outdated at this point so it would be worth trying the latest version of the library.

**Comment by tomtaila** [comment]
FWIW I was on v 1.24.2 and was getting this issue, updated to 1.31.1 no longer presents this issue.

**Comment by emu-de** [comment]
> 

Still encountering this issue on Android 11, with version 1.31.1 of the aws libraries.

**Comment by eeatonaws** [comment]
@emu-de we released a fix for this issue in version 2.38.0 of the AWS SDK for Android. Amplify Android version 1.31.1 uses an older version of the AWS SDK for Android (2.37.1). Please use version 1.31.3 of Amplify Android instead, which uses the newest version of the AWS SDK for Android with the fix for this issue.
